source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
+  if (container->ref_count_ <= 0) {
+    delete container;
+  }
+}
+
+grpc::ByteBuffer Common::makeByteBuffer(Buffer::InstancePtr bufferInstance) {
+  if (!bufferInstance) {
+    return {};
+  }
+  Buffer::RawSlice oneRawSlice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int nSlices = bufferInstance->getRawSlices(&oneRawSlice, 1);
+  if (nSlices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{nSlices, std::move(bufferInstance)};
+  if (nSlices == 1) {
+    grpc::Slice oneSlice(oneRawSlice.mem_, oneRawSlice.len_, &derefBufferInstanceContainer,
+                         container);
+    return {&oneSlice, 1};
+  }
+  STACK_ARRAY(manyRawSlices, Buffer::RawSlice, nSlices);
+  container->buffer_->getRawSlices(manyRawSlices.begin(), nSlices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(nSlices);
+  for (int i = 0; i < nSlices; i++) {
+    slices.emplace_back(manyRawSlices[i].mem_, manyRawSlices[i].len_, &derefBufferInstanceContainer,
+                        container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments); }
+  std::atomic<int> ref_count_;
+  Buffer::BufferFragmentImpl* fragments = nullptr;
+  std::vector<grpc::Slice> slices_;
+};
+
+Buffer::InstancePtr Common::makeBufferInstance(const grpc::ByteBuffer& byteBuffer) {
+  auto buffer = std::make_unique<Buffer::OwnedImpl>();
+  if (byteBuffer.Length() == 0) {
+    return buffer;
+  }
+  // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
+  // lifetime of the Slice(s) exceeds our Buffer::Instance.
+  std::vector<grpc::Slice> slices;
+  byteBuffer.Dump(&slices);
+  if (slices.size() == 0) {
+    return buffer;
+  }
+  auto container = new ByteBufferContainer(static_cast<int>(slices.size()));
+  std::function<void(const void*, size_t, const Buffer::BufferFragmentImpl*)> releaser =
+      [container](const void*, size_t, const Buffer::BufferFragmentImpl*) {
+        container->ref_count_--;
+        if (container->ref_count_ <= 0) {
+          delete container;
+        }
+      };
+  // NB: addBufferFragment takes a pointer alias to the BufferFragmentImpl which is passed in so we
+  // need to ensure that the lifetime of those objects exceeds that of the Buffer::Instance.
+  container->fragments = static_cast<Buffer::BufferFragmentImpl*>(
+      ::malloc(sizeof(Buffer::BufferFragmentImpl) * slices.size()));
+  for (size_t i = 0; i < slices.size(); i++) {
+    new (&container->fragments[i])
+        Buffer::BufferFragmentImpl(slices[i].begin(), slices[i].size(), releaser);
+  }
+  for (size_t i = 0; i < slices.size(); i++) {
+    buffer->addBufferFragment(container->fragments[i]);
+  }
+  container->slices_ = std::move(slices);
+  return buffer;
+}
+
+void Common::PrependGrpcFrameHeader(Buffer::Instance* buffer) {
```

Comment:
nit: `prependGrpcFrameHeader`
parameter should be a lvalue reference.

Reply:
Sorry, that is a tough style for me to remember as it doesn't leave a signifier at the call site that the argument is potentially changed and so has been prohibited in, well essentially every project I have ever worked on.

Also if I make it Buffer::InstancePtr& and I am holding a std::unique_ptr<OnwnedImpl> then
I have to cast to a Buffer::InstancePtr which seems worse (this occurs in common_test.cc).

If I make it Buffer::Instance& then I have to pass that by doing *buffer which seems very odd.

Which one should I do?  BufferInstancePtr& with a cast at the call site or BufferInstance& with a * at the call site?

Reply:
We note that in https://github.com/envoyproxy/envoy/blob/master/STYLE.md
> References are always preferred over pointers when the reference cannot be null. This includes both const and non-const references.

so it should be `Buffer::Instance&` with a * at call sites

Reply:
Thanx.  Will do.